src/main/java/org/jenkinsci/plugins/workflow/steps/RetryStepExecution.java

```diff
@@ -61,6 +65,18 @@ public void onSuccess(StepContext context, Object result) {
         @Override
         public void onFailure(StepContext context, Throwable t) {
             try {
+                Run run = context.get(Run.class);
+                if (run != null && t instanceof FlowInterruptedException) {
+                    InterruptedBuildAction action = run.getAction(InterruptedBuildAction.class);
+                    if (action != null) {
+                        for (CauseOfInterruption cause : action.getCauses()) {
```

Comment:
I suspect [this](https://github.com/jenkinsci/workflow-job-plugin/blob/139ac7e5da1e03ec16be698a698a3574aa748722/src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java#L273) needs to call `getCausesOfInterruption()`. While it already does call `recordCauseOfInterruption`, which records information in the `InterruptedBuildAction`, this does not allow the causes to be inspected by catching code, which is no good.

The call to `recordCauseOfInterruption` could also be removed, since `FlowInterruptedException.handle` would already add it. On the other hand, for a _caught_ exception we do currently record interruption cause, which is possibly useful especially if several people attempt to interrupt it at different times. So the question is whether we can record it for both APIs without printing duplicate messages to the log. Needs to be verified in `workflow-job`.

BTW `Executor` only records a `UserInterruption` if the interrupt is sent by a real user, not `SYSTEM`, so I am curious how your test could pass given that it does not `impersonate` anyone.